File: pwbtable/table.go
package pwbtable

import (
	"encoding/json"
	"fmt"

	"github.com/thoas/go-funk"
	"github.com/weedbox/pokerface"
)

type TableStateStatus string

const (
	TableStateStatus_TableCreated   TableStateStatus = "table_created"
	TableStateStatus_TablePausing   TableStateStatus = "table_pausing"
	TableStateStatus_TableRestoring TableStateStatus = "table_restoring"
	TableStateStatus_TableBalancing TableStateStatus = "table_balancing"
	TableStateStatus_TableClosed    TableStateStatus = "table_closed"

	TableStateStatus_TableGameOpened  TableStateStatus = "table_game_opened"
	TableStateStatus_TableGamePlaying TableStateStatus = "table_game_playing"
	TableStateStatus_TableGameSettled TableStateStatus = "table_game_settled"
	TableStateStatus_TableGameStandby TableStateStatus = "table_game_standby"
)

type Table struct {
	UpdateSerial int64       `json:"update_serial"`
	ID           string      `json:"id"`
	Meta         TableMeta   `json:"meta"`
	State        *TableState `json:"state"`
	UpdateAt     int64       `json:"update_at"`
}

type TableMeta struct {
	CompetitionID       string `json:"competition_id"`
	Rule                string `json:"rule"`
	Mode                string `json:"mode"`
	MaxDuration         int    `json:"max_duration"`
	TableMaxSeatCount   int    `json:"table_max_seat_count"`
	TableMinPlayerCount int    `json:"table_min_player_count"`
	MinChipUnit         int64  `json:"min_chip_unit"`
	ActionTime          int    `json:"action_time"`
}

type TableState struct {
	Status            TableStateStatus     `json:"status"`
	StartAt           int64                `json:"start_at"`
	SeatMap           []int                `json:"seat_map"`
	BlindState        *TableBlindState     `json:"blind_state"`
	CurrentDealerSeat int                  `json:"current_dealer_seat"`
	CurrentBBSeat     int                  `json:"current_bb_seat"`
	PlayerStates      []*TablePlayerState  `json:"player_states"`
	GameCount         int                  `json:"game_count"`
	GamePlayerIndexes []int                `json:"game_player_indexes"`
	GameState         *pokerface.GameState `json:"game_state"`
}

type TablePlayerGameAction struct {
	CompetitionID    string   `json:"competition_id"`
	TableID          string   `json:"table_id"`
	GameID           string   `json:"game_id"`
	GameCount        int      `json:"game_count"`
	Round            string   `json:"round"`
	UpdateAt         int64    `json:"update_at"`
	PlayerID         string   `json:"player_id"`
	Seat             int      `json:"seat"`
	Positions        []string `json:"positions"`
	Action           string   `json:"action"`
	Chips            int64    `json:"chips"`
	Bankroll         int64    `json:"bankroll"`
	InitialStackSize int64    `json:"initial_stack_size"`
	StackSize        int64    `json:"stack_size"`
	Pot              int64    `json:"pot"`
	Wager            int64    `json:"wager"`
}

type TablePlayerState struct {
	PlayerID          string                    `json:"player_id"`
	Seat              int                       `json:"seat"`
	Positions         []string                  `json:"positions"`
	IsParticipated    bool                      `json:"is_participated"`
	IsBetweenDealerBB bool                      `json:"is_between_dealer_bb"`
	Bankroll          int64                     `json:"bankroll"`
	IsIn              bool                      `json:"is_in"`
	GameStatistics    TablePlayerGameStatistics `json:"game_statistics"`
}

type TablePlayerGameStatistics struct {
	ActionTimes int    `json:"action_times"`
	RaiseTimes  int    `json:"raise_times"`
	CallTimes   int    `json:"call_times"`
	CheckTimes  int    `json:"check_times"`
	IsFold      bool   `json:"is_fold"`
	FoldRound   string `json:"fold_round"`
}

type TableBlindState struct {
	Level  int   `json:"level"`
	Ante   int64 `json:"ante"`
	Dealer int64 `json:"dealer"`
	SB     int64 `json:"sb"`
	BB     int64 `json:"bb"`
}

func (t Table) Clone() (*Table, error) {
	encoded, err := json.Marshal(t)
	if err != nil {
		return nil, err
	}

	var cloneTable Table
	if err := json.Unmarshal(encoded, &cloneTable); err != nil {
		return nil, err
	}

	return &cloneTable, nil
}

func (t Table) GetJSON() (string, error) {
	encoded, err := json.Marshal(t)
	if err != nil {
		return "", err
	}
	return string(encoded), nil
}

func (t Table) GetGameStateJSON() (string, error) {
	encoded, err := json.Marshal(t.State.GameState)
	if err != nil {
		return "", err
	}
	return string(encoded), nil
}

func (t Table) ParticipatedPlayers() []*TablePlayerState {
	return funk.Filter(t.State.PlayerStates, func(player *TablePlayerState) bool {
		return player.IsParticipated
	}).([]*TablePlayerState)
}

func (t Table) AlivePlayers() []*TablePlayerState {
	return funk.Filter(t.State.PlayerStates, func(player *TablePlayerState) bool {
		return player.Bankroll > 0
	}).([]*TablePlayerState)
}

func (t Table) GamePlayerIndex(playerID string) int {
	targetPlayerIdx := UnsetValue
	for idx, player := range t.State.PlayerStates {
		if player.PlayerID == playerID {
			targetPlayerIdx = idx
			break
		}
	}

	if targetPlayerIdx == UnsetValue {
		return UnsetValue
	}

	for gamePlayerIndex, playerIndex := range t.State.GamePlayerIndexes {
		if targetPlayerIdx == playerIndex {
			return gamePlayerIndex
		}
	}
	return UnsetValue
}

func (t Table) FindGamePlayerIdx(playerID string) int {
	for gamePlayerIdx, playerIdx := range t.State.GamePlayerIndexes {
		if playerIdx >= len(t.State.PlayerStates) {
			fmt.Printf("[DEBUG#FindGamePlayerIdx] TableID: %s, PlayerID: %s, GamePlayerIndexes: %+v, len(PlayerStates): %d, TableSerial: %d\n",
				t.ID,
				playerID,
				t.State.GamePlayerIndexes,
				len(t.State.PlayerStates),
				t.UpdateSerial,
			)
			continue
		}
		player := t.State.PlayerStates[playerIdx]
		if player.PlayerID == playerID {
			return gamePlayerIdx
		}
	}
	return UnsetValue
}

func (t Table) FindPlayerIdx(playerID string) int {
	for idx, player := range t.State.PlayerStates {
		if player.PlayerID == playerID {
			return idx
		}
	}
	return UnsetValue
}

func (t Table) PlayerSeatMap() map[string]int {
	playerSeatMap := make(map[string]int)
	for _, player := range t.State.PlayerStates {
		playerSeatMap[player.PlayerID] = player.Seat
	}
	return playerSeatMap
}

func (t Table) ShouldPause() bool {
	return t.State.BlindState.IsBreaking() || len(t.AlivePlayers()) < t.Meta.TableMinPlayerCount
}

func (bs TableBlindState) IsBreaking() bool {
	return bs.Level == -1
}

func (bs TableBlindState) IsSet() bool {
	return bs.Level != 0 && bs.Ante == UnsetValue && bs.Dealer == UnsetValue && bs.SB == UnsetValue && bs.BB == UnsetValue
}

File: pwbunity/Assets/Code/Model/Game/NotificationEvent/Table.cs
using System.Collections.Generic;
using JetBrains.Annotations;
using Newtonsoft.Json;

namespace Code.Model.Game.NotificationEvent
{
    public class Table
    {
        [JsonProperty("update_serial")] public long UpdateSerial; // 更新序列號 (數字越大越晚發生)
        [JsonProperty("id")] public string ID; // 桌次 Unique ID
        [JsonProperty("meta")] public TableMeta Meta; // 桌次固定資料
        [JsonProperty("state")] [CanBeNull] public TableState State; // 桌次動態資料
        [JsonProperty("update_at")] public long UpdateAt; // 更新時間 (Seconds)
    }

    public class TableMeta
    {
        [JsonProperty("competition_id")] public string CompetitionID; // 賽事 ID
        [JsonProperty("rule")] public string Rule; // 德州撲克規則, 常牌(default), 短牌(short_deck), 奧瑪哈(omaha)
        [JsonProperty("mode")] public string Mode; // 賽事模式 (CT, MTT, Cash)
        [JsonProperty("max_duration")] public int MaxDuration; // 比賽時間總長 (Seconds)
        [JsonProperty("table_max_seat_count")] public int TableMaxSeatCount; // 每桌人數上限
        [JsonProperty("table_min_player_count")] public int TableMinPlayerCount; // 每桌最小開打數
        [JsonProperty("min_chip_unit")] public long MinChipUnit; // 最小單位籌碼量
        [JsonProperty("action_time")] public int ActionTime; // 玩家動作思考時間 (Seconds)
    }

    public class TableState
    {
        [JsonProperty("status")] public string Status; // 當前桌次狀態
        [JsonProperty("start_at")] public long StartAt; // 開打時間 (Seconds)
        [JsonProperty("seat_map")] [CanBeNull] public List<int> SeatMap; // 座位入座狀況，index: seat index (0-8), value: TablePlayerState index (-1 by default)
        [JsonProperty("blind_state")] [CanBeNull] public TableBlindState BlindState; // 盲注狀態
        [JsonProperty("current_dealer_seat")] public int CurrentDealerSeat; // 當前 Dealer 座位編號
        [JsonProperty("current_bb_seat")] public int CurrentBbSeat; // 當前 BB 座位編號
        [JsonProperty("player_states")] [CanBeNull] public List<TablePlayerState> PlayerStates; // 賽局桌上玩家狀態
        [JsonProperty("game_count")] public int GameCount; // 執行牌局遊戲次數 (遊戲跑幾輪)
        [JsonProperty("game_player_indexes")] [CanBeNull] public List<int> GamePlayerIndexes; // 本手正在玩的 PlayerIndex 陣列 (陣列 index 為從 Dealer 位置開始的 PlayerIndex)，GameEngine 用
        [JsonProperty("game_state")] [CanBeNull] public GameState GameState; // 本手狀態
        [JsonProperty("seat_changes")] [CanBeNull] public TableGameSeatChanges SeatChanges; // 新的一手座位狀況
        [JsonProperty("last_player_game_action")] [CanBeNull] public TablePlayerGameAction LastPlayerGameAction; // 最新一筆玩家牌局動作
    }

    public class TableBlindState
    {
        [JsonProperty("level")] public int Level; // 盲注等級(-1 表示中場休息)
        [JsonProperty("ante")] public long Ante; // 前注籌碼量
        [JsonProperty("dealer")] public long Dealer; // 庄位籌碼量
        [JsonProperty("sb")] public long Sb; // 大盲籌碼量
        [JsonProperty("bb")] public long Bb; // 小盲籌碼量
    }

    public class TablePlayerState
    {
        [JsonProperty("player_id")] public string PlayerID; // 玩家 ID
        [JsonProperty("seat")] public int Seat; // 座位編號 0 ~ 8
        [JsonProperty("positions")] [CanBeNull] public List<string> Positions; // 場上位置
        [JsonProperty("is_participated")] public bool IsParticipated; // 玩家是否參戰
        [JsonProperty("is_between_dealer_bb")] public bool IsBetweenDealerBb; // 玩家入場時是否在 Dealer & BB 之間
        [JsonProperty("bankroll")] public long Bankroll; // 玩家身上籌碼
        [JsonProperty("is_in")] public bool IsIn; // 玩家是否入座
        [JsonProperty("game_statistics")] public GameStatistics GameStatistics; // 玩家每手遊戲統計
    }
    
    public class GameStatistics
    {
        [JsonProperty("action_times")] public int ActionTimes; // 下注動作總次數
        [JsonProperty("raise_times")] public int RaiseTimes; // 加注總次數
        [JsonProperty("call_times")] public int CallTimes; // 跟注總次數
        [JsonProperty("check_times")] public int CheckTimes; // 過牌總次數
        [JsonProperty("is_fold")] public bool IsFold; // 是否蓋牌
        [JsonProperty("fold_round")] public string FoldRound; // 蓋牌回合
    }

    public class TableGameSeatChanges
    {
        [JsonProperty("new_dealer")] public int NewDealer; // 下一手 Dealer 座位編號
        [JsonProperty("new_sb")] public int NewSb; // 下一手 SB 座位編號
        [JsonProperty("new_bb")] public int NewBb; // 下一手 BB 座位編號
    }

    public class TablePlayerGameAction
    {
        [JsonProperty("table_id")] public string TableID; // 桌次 ID
        [JsonProperty("game_id")] public string GameID; // 遊戲 ID
        [JsonProperty("game_count")] public int GameCount; // 執行牌局遊戲次數 (遊戲跑幾輪)
        [JsonProperty("round")] public string Round; // 哪回合
        [JsonProperty("update_at")] public long UpdateAt; // 更新時間 (Seconds)
        [JsonProperty("player_id")] public string PlayerID; // 玩家 ID
        [JsonProperty("seat")] public int Seat; // 座位編號 0 ~ 8
        [JsonProperty("positions")] [CanBeNull] public List<string> Positions; // 場上位置
        [JsonProperty("action")] public string Action; // 動作
        [JsonProperty("chips")] public long Chips; // 下注籌碼量
    }
}